### PR TITLE
[WIP] Native pub/sub support in MSMQ transport

### PR DESCRIPTION
**Pull request by SzymonPobiega**
The native support for pub/sub in MSMQ based on a built-in subscription store
 * SQL Server subscription store (built-in)
 * RavenDB subscription store (https://github.com/SzymonPobiega/NServiceBus.Msmq.RavenDB) (separate package)

There is a similar SPIKE/WIP for SQL Server transport is here: https://github.com/Particular/NServiceBus.SqlServer/pull/168

### CODE REVIEW

**Comment by andreasohlund** [comment]
:heart: it, I did the same for the devtransport this weekend https://github.com/Particular/NServiceBus/pull/2950

File: src/NServiceBus.AcceptanceTests/ConfigureMsmqTransport.cs [hunk 1]
```diff
@@ -12,22 +13,35 @@
...
+        configuration.UseTransport<MsmqTransport>()
+            .ConnectionString(settings["Transport.ConnectionString"])
+            .UseSubscriptionStore<SqlServerSubscriptionStore>().ConnectionString(connectionString);
         return Task.FromResult(0);
     }
 
     public Task Cleanup()
     {
+        //Clean subscription store
+        using (var connection = new SqlConnection(connectionString))
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/ConfigureMsmqTransport.cs]: I'm disappointed that you didn't use the async API ;)

File: src/NServiceBus.Core.Tests/Config/When_no_custom_configuration_source_is_specified.cs [hunk 2]
```diff
@@ -14,6 +15,7 @@ public class When_no_custom_configuration_source_is_specified
             var config = new EndpointConfiguration();
 
             config.SendOnly();
+            config.UseTransport<MsmqTransport>().UseSubscriptionStore<FakeSubscriptionStore>();
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Config/When_no_custom_configuration_source_is_specified.cs]: I think we should make this a mandatory setting.

Comment (SzymonPobiega) [inline src/NServiceBus.Core.Tests/Config/When_no_custom_configuration_source_is_specified.cs]: Make sense, but how? Without introducing an additional extension method?

**Comment by bording** [comment]
If you still have to configure a separate database to get this working, what's the benefit of doing this vs. the existing message-based subscriptions via persistence? It looks like `MsmqMessageDispatcher` still has to send separate messages out to every endpoint for a Publish.

**Comment by distantcam** [comment]
Sorry, I'm putting my squad/project manager hat on.

- Is this a bug?
- Is this part of v6 work?
- Has this been approved?
- Who is the taskforce for this work?
- Why has it suddenly gone into progress?

**Comment by andreasohlund** [comment]
Not going to speak for @SzymonPobiega but I'm pretty sure that this is a spike he did last weekend to show that we should try to move pub/sub into each transport instead of having "message driven pubsub" as a separate "component". @SzymonPobiega ? (hence the `[Spike]` prefix)

Should we stop waffle from making pulls automatically go into inprogress?

**Comment by SzymonPobiega** [comment]
@distantcam I was bored so I wanted to experiment. I should probably do it in my fork but then it would be much less visible.

**Comment by distantcam** [comment]
@SzymonPobiega No problem. Even on your own fork once you create a PR  to discuss it Waffle will automatically add the `State: In Progress` label, which is the problem.

Not sure how we'll handle that in the future, but I'll just remove the label for now.

**Comment by andreasohlund** [comment]
Labeling as `Type: Spike` and prefixing with [WIP]

**Comment by andreasohlund** [comment]
@SzymonPobiega can you flesh out the title (mention MSMQ etc)

**Comment by SzymonPobiega** [comment]
@andreasohlund do you think this should be paused at this stage or it there anything I can do now?

**Comment by danielmarbach** [comment]
@SzymonPobiega you could raise an issue in PD or here. Flesh out everything in the issue and then ping the squad.

**Comment by danielmarbach** [comment]
@SzymonPobiega can we close this PR and capture the ideas/outcome in a PD issue?

**Comment by SzymonPobiega** [comment]
@danielmarbach let it live a while longer. I've just proposed it as a solution for the V3/V2 subscriber compat problem https://github.com/Particular/NServiceBus.SqlServer/issues/181

**Comment by SzymonPobiega** [comment]
Closing for now. Will re-visit later.

**Pull request closed without merging**